### PR TITLE
Create conformance test for scs-0214-v1

### PR DESCRIPTION
**Pull request by cah-hbaum**
As discussed in https://github.com/SovereignCloudStack/standards/issues/481, we need to create a conformance test for SCS-0214-v1. This PR does that and also includes a separate commit containing the necessary changes for the `Conformance Tests` section of the standard. If we don't want to include this, the commit will be split up into a separate PR.

Resolves https://github.com/SovereignCloudStack/issues/issues/477

### CODE REVIEW

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    for label in labels:
+        if len(labels[label]) <= 1:
+            if len(labels[label]) == 1:
+                logger.warning(f"There seems to be no distribution across multiple {label.split('/')[1]}s "
+                               "or labels aren't set correctly across nodes.")
+            elif len(labels[label]) == 0:
+                logger.warning(f"The label for {label.split('/')[1]}s don't seem to be set for all nodes.")
+        else:
+            logger.info(f"The nodes are distributed across {str(len(labels[label]))} {label.split('/')[1]}s.")
+            return
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:209]: The `return` here comes as a little surprise to me. Are we happy if just one of the labels has more than 1 value? Maybe this can be documented a bit (preferably in the standard itself, but at least make a little remark here).

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+    nodes = await get_k8s_cluster_labelled_nodes(config.kubeconfig)
+
+    if len(nodes) < 2:
+        logger.error("The tested cluster only contains a single node, which can't comply with the standard.")
+        return 2
+
+    labelled_master_nodes = [node for node in nodes if "node-role.kubernetes.io/control-plane" in node]
+    try:
+        if len(labelled_master_nodes) > 1 and len(nodes) > 2:
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: shouldn't this be `>= 1` and `=> 2`?

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            worker_nodes = [node for node in nodes if "node-role.kubernetes.io/control-plane" not in node]
+            # Compare the labels of both types, since we have enough of them with labels
+            compare_labels(labelled_master_nodes, "master")
+            compare_labels(worker_nodes, "worker")
+        else:
+            compare_labels(nodes)
+    except DistributionException:
+        logger.error("The distribution of nodes described in the standard couldn't be detected.")
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:244]: Preferably pass this string literal as argument to the exception upon throwing it; here just use `str(exc)` or something similar.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        for key in labels:
+            labels[key].add(node[key])
+
+    for label in labels:
+        if len(labels[label]) <= 1:
+            if len(labels[label]) == 1:
+                logger.warning(f"There seems to be no distribution across multiple {label.split('/')[1]}s "
+                               "or labels aren't set correctly across nodes.")
+            elif len(labels[label]) == 0:
+                logger.warning(f"The label for {label.split('/')[1]}s don't seem to be set for all nodes.")
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:206]: Should we produce this warning also when there are more than 1 nodes with the label, but fewer than total nodes?

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+import asyncio
+import getopt
+import kubernetes_asyncio
+import logging
+import logging.config
+import sys
+import yaml
+
+
+logging_config = {
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:43]: This seems kinda redundant given that the same info is in the yaml file?

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+The following return values are possible:
+    0 - Distribution according to the standard could be verified.
+    1 - The help output was delivered.
+    2 - No distribution according to the standard could be detected for the nodes available.
+
+The following arguments can be set:
+    -c/--config PATH/TO/CONFIG         - Path to the config file of the test script
+    -k/--kubeconfig PATH/TO/KUBECONFIG - Path to the kubeconfig of the server we want to check
+    -h                                 - Output help
+    """)
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:103]:
```suggestion
""")
```

**Review by mbuechse — COMMENTED** [review]
A few things that I think should be addressed. If you are time-constrained, I offer to assist.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    return nodes
+
+
+def compare_labels(node_list, node_type="master"):
+
+    labels = {
+        "topology.kubernetes.io/region": set(),
+        "topology.kubernetes.io/zone": set(),
+        "kubernetes.io/hostname": set(),
+    }
```
[the remark this answers is not in the document]
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:194]: Yes its intentional, we don't need it here.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 1 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:209]: Will document it.

Just to clarify it here as well, we're fine with one of these labels.
Basically, if we see that nodes are in different regions, it should (if the labeling is done correctly) not matter, if the (availability) zone or hostname is the same, since the same zone name or hostname could be used in different regions. Same thing goes for availability zones and hostnames.

If we don't want/can't rely on that, what would our way to go be? Throwing an error would probably invalidate quite a few clusters that are configured correctly (and some that are just labelled wrong, but that could always be the case).

I think being a bit more generous instead of too strict (also with regard to our time limit on this) is the better way.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 2 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: No, since if there is only 1 master node, we can't compare it with another master node. And we can also only do this, if we have more than 2 nodes in general, since this would mean we have at least one worker node.

I could argue about `> 2` being `>= 2`, since there could be a cluster without worker nodes and we would do these edge cases in different `if else` paths right now.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 3 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:244]: Understandable, done in the fixup.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 4 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:206]: Yes seems like a good idea.
I will include that in the fixup.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 5 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:43]: This is there if a `yaml` file isn't provided.

Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:43]: I know. Then it should just be provided (and the fallback can be way more minimal)

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 2 above)
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: My understanding was that once there is at least one occurrence of "node-role.kubernetes.io/control-plane", then we know that these labels are set, and we can distinguish between master nodes and worker nodes; otherwise we just assume that every node is a master node. Can a cluster with only one master node pass the test at all?

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 1 above)
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:209]: If I understand your answer correctly, we should check the labels in a certain order, right? In this case, region first, then zone, then hostname.

Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:209]: This is not clear from the code. Yes, that is the order in which they are named when the dict `labels` is constructed, but you rely on an implementation detail, namely that this order is preserved when iterating over the dict. You have to document this assumption, or better yet, iterate over a tuple. At any rate, it should also be document why the order is important.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:194]: I understood this now; the control-plane label is only for distinguishing node types. Good to know.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 2 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: > Can a cluster with only one master node pass the test at all?
I would personally say no, but not because the setup is wrong, but because such a cluster wouldn't be constructed for availability or long-lasting services, but its probably there for development, testing, etc.
So its probably not something you would test for that.

Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: My question was meant with respect to the standard, not your opinion ;)

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 1 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:209]: I will document it in the text as well (and the tuple is coming)

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 2 above)
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: If you change the condition to `>= 1`, then such a setup would fail the test (because `compare_labels` wouldn't reach the return statement, but the raise statement), and if the standard agrees with that result, then I don't see why we need `> 1`.

Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: The standard says that multiple control plane nodes should be distributed. So it would pass.

Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:236]: Got it.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    """Get the labels of the k8s cluster under test."""
+    _ = await kubernetes_asyncio.config.load_kube_config(kubeconfig)
+
+    nodes = list()
+    labels = [
+        "kubernetes.io/hostname",
+        "topology.kubernetes.io/zone",
+        "topology.kubernetes.io/region",
+        "node-role.kubernetes.io/control-plane"
+    ]
```
[the remark this answers is not in the document]
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:169]: Added a tuple and put it into both functions as a parameter.

**Comment by cah-hbaum** [comment]
Added a fixup containing the mentioned changes. I hope I didn't forget anything @mbuechse

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+    # infrastructure parts to smaller ones. So we first look if nodes are distributed
+    # across regions, then zones and then hosts. If one of these requirements is fulfilled,
+    # we probably don't need to check anymore, since a distribution was already detected.
+    labels = (
+        "topology.kubernetes.io/region",
+        "topology.kubernetes.io/zone",
+        "kubernetes.io/hostname",
+    )
+
+    nodes = await get_k8s_cluster_labelled_nodes(config.kubeconfig, labels + tuple("node-role.kubernetes.io/control-plane"))
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:234]:
```suggestion
    nodes = await get_k8s_cluster_labelled_nodes(config.kubeconfig, labels + ("node-role.kubernetes.io/control-plane", ))
```

Your version does strange things:

```console
$ ipython3
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.31.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: tuple("node-role.kubernetes.io/control-plane")
Out[1]: 
('n',
 'o',
 'd',
 'e',
 '-',
 'r',
 'o',
 'l',
 'e',
 '.',
 'k',
 'u',
 'b',
 'e',
 'r',
 'n',
 'e',
 't',
 'e',
 's',
 '.',
 'i',
 'o',
 '/',
 'c',
 'o',
 'n',
 't',
 'r',
 'o',
 'l',
 '-',
 'p',
 'l',
 'a',
 'n',
 'e')
```

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+                    # Node is not labelled as a control plane node
+                    pass
+            nodes.append(node)
+
+    return nodes
+
+
+def compare_labels(node_list, labels, node_type="master"):
+
+    label_data = dict()
```
Comment (mbuechse) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:184]:
```suggestion
    label_data = {key: set() for key in labels}
```

Then lines 188, 189 can be removed.

Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py]: Updated.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+    for node in node_list:
+        for key in labels:
+            if key not in label_data:
+                label_data[key] = set()
+            try:
+                label_data[key].add(node[key])
+            except KeyError:
+                logger.warning(f"The label for {labels[key].split('/')[1]}s don't seem to be set for all nodes.")
```
[the remark this answers is not in the document]
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py]: Updated.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 9 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py]: Interesting, thank you.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    for label in labels:
+        if len(labels[label]) <= 1:
+            if len(labels[label]) == 1:
+                logger.warning(f"There seems to be no distribution across multiple {label.split('/')[1]}s "
+                               "or labels aren't set correctly across nodes.")
+            elif len(labels[label]) == 0:
+                logger.warning(f"The label for {label.split('/')[1]}s don't seem to be set for all nodes.")
+        else:
+            logger.info(f"The nodes are distributed across {str(len(labels[label]))} {label.split('/')[1]}s.")
+            return
```
[the remark this answers is not in the document]
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py]: Yes the `labels` needed to be replaced with `label_data`, already did that now.

I'm just confused about the second part, which comment about these lines? Or did these already switch line?

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 4 above)
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py]: This comment here. You are right, the line numbers have changed, it doesn't get shown any more in the files view.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py]: I replied to the comment in question.

[the remark this answers is not in the document]
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py]: Ah, you're right, of course! We could use a `collections.Counter` instead of a set; this should cover both use cases. But I think we can postpone this. I don't know how important this warning is anyway.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 5 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:43]: Its already more minimal than the config, but maybe it can be minimized more(?) Not sure though.
I fixed some small artifacts I found in the config.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+        # Otherwise, we initialize logging with the included literal
+        setup_logging(config.logging or logging_config)
+
+    if config.kubeconfig is None:
+        raise ConfigException("A kubeconfig needs to be set in order to test a k8s cluster version.")
+
+    return config
+
+
+async def get_k8s_cluster_labelled_nodes(kubeconfig, labels):
```
Comment (martinmo) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:159]: To make the intent clearer that the second argument is used for filtering, it should have name such as `interesting_labels`.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+        for item in result.items:
+            node = dict()
+            node_labels = item.metadata.labels
+            for label in labels:
+                try:
+                    node[label] = node_labels[label]
+                except KeyError:
+                    # Node is not labelled as a control plane node
+                    pass
+            nodes.append(node)
```
Comment (martinmo) [inline Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:177]:
```suggestion
        for node in result.items:
            filtered_labels = {
                label: value
                for label, value in node.metadata.labels.items()
                if label in interesting_labels
            }
            nodes.append(filtered_labels)
```

I think this is easier to read. Also, that comment was misleading ("Node is not labelled as a control plane node").

**Comment by martinmo** [comment]
> Additionally, with manual testing I could only verify that the following cases have the expected conformance test outcome

I forgot to mention that, to me, the testing logic seems to be ok. Still, it would be good to have an easy way to verify the different scenarios, without having access to different kinds of clusters.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:1]: Youre right, I will correct that.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+        if hasattr(e, 'message'):
+            logger.error(e.message)
+        print_usage()
+        return 1
+
+    # It is important to note, that the order of these labels matters for this test.
+    # Since we want to check if nodes are distributed, we want to do this from bigger
+    # infrastructure parts to smaller ones. So we first look if nodes are distributed
+    # across regions, then zones and then hosts. If one of these requirements is fulfilled,
+    # we probably don't need to check anymore, since a distribution was already detected.
```
[the remark this answers is not in the document]
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:224]: I wont do a Star Wars joke here and instead just remove that.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 13 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:159]: Sure why not.

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 14 above)
Comment (cah-hbaum) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:177]: Yeah the comment was ooolllddd.

And I'm not that "pythonic", thats why I never come up with that, but thank you for the suggestion.

**Comment by cah-hbaum** [comment]
> > Additionally, with manual testing I could only verify that the following cases have the expected conformance test outcome
> 
> I forgot to mention that, to me, the testing logic seems to be ok. Still, it would be good to have an easy way to verify the different scenarios, without having access to different kinds of clusters.

I agree with you on this, we can probably look into this later on. (Sadly not until tomorrow).

I also fixed/updated/changed most of the things you mentioned in the other comments, so you can recheck tomorrow =)

File: Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py (hunk 5 above)
Comment (mbuechse) [reply Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py:43]: Let's come back to this point later. It's not mission critical.

**Review by mbuechse — APPROVED** [review]

**Review by martinmo — APPROVED** [review]
LGTM & we agreed via chat to add tests for the tests later.
